### PR TITLE
Streamlining action selection and route values

### PR DESCRIPTION
**Pull request by rynowak**
Removes a few concepts we don't need
Improves documentation
More things are IDictionary instead of other random types

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs [hunk 1]
```diff
@@ -119,16 +119,14 @@ public static IList<ControllerActionDescriptor> Build(ApplicationModel applicati
...
-                            string.Empty));
+                        actionDescriptor.RouteValues.Add(TreeRouter.RouteGroupKey, string.Empty);
                     }
 
                     // Add a route constraint with DenyKey for each constraint in the set to all the
                     // actions that don't have that constraint. For example, if a controller defines
                     // an area constraint, all actions that don't belong to an area must have a route
                     // constraint that prevents them from matching an incoming request.
-                    AddRemovalConstraints(actionDescriptor, removalConstraints);
+                    AddRemovalConstraints(actionDescriptor, routeValueKeys);
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs]: `AddRouteValue` - also the comment above needs to be updated. `DenyKey` no longer exists.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs [hunk 2]
```diff
@@ -459,72 +456,48 @@ private static ParameterDescriptor CreateParameterDescriptor(PropertyModel prope
...
             {
-                if (constraintAttribute.BlockNonAttributedActions)
-                {
-                    removalConstraints.Add(constraintAttribute.RouteKey);
-                }
+                keys.Add(kvp.Key);
 
                 // Skip duplicates - this also means that a value on the action will take precedence
-                if (!HasConstraint(actionDescriptor.RouteConstraints, constraintAttribute.RouteKey))
+                if (!actionDescriptor.RouteValues.ContainsKey(kvp.Key))
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs]: Can you move this into `AddIfNotKeys(string key, string value)`? We repeat this several times in this type.

File: src/Microsoft.AspNetCore.Mvc.Core/Routing/KnownRouteValueConstraint.cs [hunk 3]
```diff
@@ -71,18 +72,19 @@ private string[] GetAndCacheAllMatchingValues(string routeKey, HttpContext httpC
...
-                        .Select(ad => ad.RouteConstraints.FirstOrDefault(
-                            c => c.RouteKey == routeKey &&
-                            c.KeyHandling == RouteKeyHandling.RequireKey))
-                        .Where(rc => rc != null)
-                        .Select(rc => rc.RouteValue)
-                        .Distinct()
-                        .ToArray();
-
-                valuesCollection = new RouteValuesCollection(version, routeValueCollection);
+                var values = new HashSet<string>();
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/Routing/KnownRouteValueConstraint.cs]: `StringComparer.Ordinal`

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/AttributeRoute.cs [hunk 4]
```diff
@@ -243,11 +243,10 @@ private static IEnumerable<RouteInfo> GroupRouteInfosByGroupId(List<RouteInfo> r
             Dictionary<string, RouteTemplate> templateCache,
             ActionDescriptor action)
         {
-            var constraint = action.RouteConstraints
-                .FirstOrDefault(c => c.RouteKey == TreeRouter.RouteGroupKey);
-            if (constraint == null ||
-                constraint.KeyHandling != RouteKeyHandling.RequireKey ||
-                constraint.RouteValue == null)
+            string value;
+            action.RouteValues.TryGetValue(TreeRouter.RouteGroupKey, out value);
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/AttributeRoute.cs]: `if (!action.RouteValues.TryGetValue(TreeRouter.RouteGroupKey, out value) && string.IsNullOrEmpty(value))`

File: test/Microsoft.AspNetCore.Mvc.Razor.Test/RazorViewEngineTest.cs [hunk 5]
```diff
@@ -1495,17 +1495,12 @@ public void ViewLocationFormats_ContainsExpectedLocations()
         }
 
         [Fact]
-        public void GetNormalizedRouteValue_ReturnsValueFromRouteConstraints_IfKeyHandlingIsRequired()
+        public void GetNormalizedRouteValue_ReturnsValueFromRouteValues_IfKeyHandlingIsRequired()
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Razor.Test/RazorViewEngineTest.cs]: `GetNormalizedRouteValue_ReturnsValueFromRouteValues_IfKeyExistsInRouteValues`

**Comment by pranavkm** [comment]
:shipit: once comments are addressed.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs]: I'd really prefer not.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs]: :ok:

File: src/Microsoft.AspNetCore.Mvc.Core/Routing/KnownRouteValueConstraint.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Routing/KnownRouteValueConstraint.cs]: This should actually be ignorecase

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/AttributeRoute.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/AttributeRoute.cs]: No, I don't want the `&&`

**Pull request closed without merging**